### PR TITLE
Convert rust-analyzer to an in-tree tool

### LINKED ISSUE #12815: Move rust-analyzer from a git submodule to a git subtree
(Related to: long-term proc macro support plan)

## Before (submodule)

RA is a git submodule [in src/tools](https://github.com/rust-lang/rust/tree/master/src/tools). When a "sync" happens, `rust-lang/rust` is updated to point to a different commit ([example](https://github.com/rust-lang/rust/pull/99176/files)).

The changes suggested in #12803 (use the sysroot `proc_macro` crate) make RA/rustc incompatibilities more visible in that the submodule could then stop building.

Fixing that would involve two PRs:

  * One to `rust-analyzer`, fixing `proc-macro-srv` to follow `proc_macro` bridge changes
  * One to `rust`, bumping `rust-analyzer`

If the sync cycle isn't changed, this means the rust-analyzer RA component could be missing for up to a week, or even more, if the bridge changes haven't been followed in the `rust-analyzer` repo.

## After (subtree)

Just like `clippy`, `rust-analyzer` would be a subtree (also in `src/tools`).

Feature development / bug fixes to RA happen as usual in the `rust-lang/rust-analyzer` repository. These can be synced to `rust-lang/rust` whenever convenient ([clippy->rust example PR](https://github.com/rust-lang/rust/pull/99406)).

Changes to the `proc_macro` bridge are followed in the `rust-lang/rust` repository. These can be synced to `rust-lang/rust-analyzer` whenever convenient ([rust->clippy example commit](https://github.com/rust-lang/rust-clippy/commit/f074034590bdcdfbf31114924df091017097f0fa))

**proc-macro-srv never breaks again** (in the rustup component version of rust-analyzer), because it's immediately fixed as part of any PRs to `rust-lang/rust`.

## Why does this matter?

A PR was _just merged_ to "[add support for the 1.64 ABI](https://github.com/rust-lang/rust-analyzer/pull/12795)", but there's no such thing as "the 1.64 ABI", and it's going to be instabroken, see https://github.com/rust-lang/rust-analyzer/issues/12806.

## Prior art

Similar discussion for rustfmt: https://github.com/rust-lang/rust/issues/82385 (converted to a subtree in [May 2021](https://github.com/rust-lang/rust/pull/82208))

---

Tagging:

  * @mystor who's been making proc_macro bridge changes, to confirm that `rust-lang/rust` folks are happy to keep the `proc-macro-srv` crate up-to-date with bridge changes (pinging other maintainers as needed)
  * @jonas-schievink who's been keeping the "multi ABI support" approach up-to-date by manually vendoring+tuning several copies of the `proc_macro` bridge in the rust-analyzer repo
  * @lnicola who's been doing most of the syncs I've seen in `rust-lang/rust` history
  * @jyn514 who discussed using subtrees [on Zulip](https://rust-lang.zulipchat.com/#narrow/stream/185405-t-compiler.2Frust-analyzer/topic/rust-analyzer.2312803/near/290046728)
  * @flodiebold, @Veykril, @bjorn3 who interacted with #12803 
  * @oli-obk who offered to connect with clippy folks in case RA folks have any questions about git subtree
  * @cuviper who started work on [getting RA tests to run on rust CI](https://github.com/rust-lang/rust/pull/99444)

### PR DESCRIPTION
**Pull request by fasterthanlime**
This re-adds `rust-lang/rust-analyzer` as a git subtree rather than a submodule.

Closes https://github.com/rust-lang/rust-analyzer/issues/12815.

Prior attempt (research PR): https://github.com/rust-lang/rust/pull/99465

  * [x] Remove submodule: `git rm -f src/tools/rust-analyzer`
  * [x] Add subtree: `git subtree add -P src/tools/rust-analyzer https://github.com/rust-lang/rust-analyzer.git master`
  * [x] Move to `SourceType::InTree`,
  * [x] Enable `rust-analyzer/in-rust-tree` feature when built through `x.py`
  * [x] Add 'check' step
  * [x] Add 'test' step

With this PR, rust-analyzer becomes an "in-tree" tool. Syncs can happen in both directions, see [clippy's relevant book section](https://doc.rust-lang.org/nightly/clippy/development/infrastructure/sync.html).

Making sure `proc-macro-srv` doesn't break when the proc_macro bridge changes effectively becomes the responsibility of `rust-lang/rust` contributors. These days, that's mostly @mystor, who has been consulted throughout the process. I'm also making myself available in case there's questions / work needed that nobody else signed up for.

This doesn't change rust-analyzer's release cycle. After this PR is merged and the next nightly goes out, one can point `rust-analyzer.procMacro.server` to the rustup-provided `rust-analyzer` binary. Changes to improve the situation further (auto-discovery/install of the rust-analyzer component) will happen in `rust-lang/rust-analyzer` and be synced here eventually.

### CODE REVIEW

**Comment by fasterthanlime** [comment]
@jyn514 I'm stuck on this when running `x.py check rust-analyzer`:

```
thread 'main' panicked at 'no output generated for "libgoto_def-501a4807e737f416" "rmeta"', compile.rs:1448:21
stack backtrace:
   0: rust_begin_unwind
             at /rustc/efd358333ac55ca50f786ca1ab841c003bfdfff0/library/std/src/panicking.rs:584:5
   1: core::panicking::panic_fmt
             at /rustc/efd358333ac55ca50f786ca1ab841c003bfdfff0/library/core/src/panicking.rs:142:14
   2: bootstrap::compile::run_cargo
             at ./src/bootstrap/compile.rs:1448:21
   3: <bootstrap::check::RustAnalyzer as bootstrap::builder::Step>::run
             at ./src/bootstrap/check.rs:353:9
   4: bootstrap::builder::Builder::ensure
             at ./src/bootstrap/builder.rs:2126:23
   5: <bootstrap::check::RustAnalyzer as bootstrap::builder::Step>::make_run
             at ./src/bootstrap/check.rs:319:9
   6: bootstrap::builder::StepDescription::maybe_run
             at ./src/bootstrap/builder.rs:270:13
   7: bootstrap::builder::StepDescription::run
             at ./src/bootstrap/builder.rs:336:17
   8: bootstrap::builder::Builder::run_step_descriptions
             at ./src/bootstrap/builder.rs:841:9
   9: bootstrap::builder::Builder::execute_cli
             at ./src/bootstrap/builder.rs:821:9
  10: bootstrap::Build::build
             at ./src/bootstrap/lib.rs:702:13
  11: bootstrap::main
             at ./src/bootstrap/bin/main.rs:50:5
  12: core::ops::function::FnOnce::call_once
             at /rustc/efd358333ac55ca50f786ca1ab841c003bfdfff0/library/core/src/ops/function.rs:248:5
note: Some details are omitted, run with `RUST_BACKTRACE=full` for a verbose backtrace.
```

Not sure what's causing this, here's my current Step definition:

https://github.com/rust-lang/rust/blob/aec26157b5a33b0ca036293e5b9b8729db50e947/src/bootstrap/check.rs#L304-L361

**Comment by fasterthanlime** [comment]
Oh wow, it comes from a rust-analyzer example:

```
studying filename /home/amos/bearcove/rust/build/x86_64-unknown-linux-gnu/stage0-tools/x86_64-unknown-linux-gnu/release/examples/libgoto_def-501a4807e737f416.rmeta
```

Passing `--tests --benches` instead of `--all-targets` ought to do the trick.

**Comment by jyn514** [comment]
You can't use cargo check in-tree, you need a custom config: https://rustc-dev-guide.rust-lang.org/building/suggested.html#configuring-rust-analyzer-for-rustc

What is the state of this PR? should I start reviewing it or are you still rapidly making changes?

**Comment by fasterthanlime** [comment]
> You can't use cargo check in-tree, you need a custom config: https://rustc-dev-guide.rust-lang.org/building/suggested.html#configuring-rust-analyzer-for-rustc

Yes, that's already with the command override going through `x.py check` (I think that's the problem - diagnostics coming out of that are relative to the subtree, but the rust-analyzer vscode expects them to be relative to the workspace root).

> What is the state of this PR? should I start reviewing it or are you still rapidly making changes?

What's there is good to review - I'm adding the test step real quick.

File: src/bootstrap/check.rs
Comment (jyn514) [inline src/bootstrap/check.rs:350]: fyi x.py clippy is a giant hack right now, I wouldn't spend too much time trying to get it working. see https://github.com/rust-lang/rust/pull/97443 for the status of making that more reliable.

File: src/bootstrap/check.rs
Comment (jyn514) [inline src/bootstrap/check.rs:354]: I know these are the same, but I would prefer to use `compiler.stage` just for my sanity, in case we ever make it configurable in the future.
```suggestion
            compiler.stage, "rust-analyzer", &compiler.host.triple, target.triple
```

File: src/bootstrap/tool.rs
Comment (jyn514) [inline src/bootstrap/tool.rs:699]: this is a strange path. normally this would be the top-level path (src/tools/rust-analyzer), not the specific crate. if you want to allow building a single crate instead of the whole workspace, you can use `run.crates("rust-analyzer")`.

(don't remember if it's crate or crates - definitely *not* `all_crates`, that does something different, sorry. trying to get rid of that function so it's less confusing.)

File: src/bootstrap/test.rs
Comment (jyn514) [inline src/bootstrap/test.rs:395]: Why are you overriding this? I think cargo will do different feature / dependency resolution for the different workspace, no?

File: src/bootstrap/tool.rs
Comment (jyn514) [inline src/bootstrap/tool.rs:695]: Why do you have ONLY_HOSTS set? that will prevent people from cross-compiling, which shouldn't be necessary. tests are usually host only but there's nothing target-specific about rust-analyzer that I know of.

File: src/bootstrap/tool.rs
Comment (jyn514) [inline src/bootstrap/tool.rs:722]: this only builds that top-level crate, is that intentional?

File: src/bootstrap/check.rs
Comment (fasterthanlime) [reply src/bootstrap/check.rs:350]: note that this branch executes if we're _not_ running clippy, ie. if we're running check.

I can make it unconditional if you prefer!

File: src/bootstrap/test.rs
Comment (jyn514) [reply src/bootstrap/test.rs:395]: this is likely also what's breaking `x.py check` in your IDE

Comment (fasterthanlime) [reply src/bootstrap/test.rs:395]: cargo doesn't set that, unfortunately: https://github.com/rust-lang/cargo/issues/3946

RA issue: https://github.com/rust-analyzer/expect-test/issues/33

File: src/bootstrap/check.rs
Comment (jyn514) [reply src/bootstrap/check.rs:350]: oh I see, this is adapted from the step above that uses `--all-targets`. ok, this is fine to keep as-is

File: src/bootstrap/test.rs
Comment (jyn514) [reply src/bootstrap/test.rs:395]: oh I see, this is only read by rust-analyzer, not cargo itself. can you add that as a comment?

**Comment by fasterthanlime** [comment]
The latest failure is due to a test trying to run rustfmt from the `stable` toolchain. That test should probably be disabled when running in Rust CI.

File: src/tools/rust-analyzer/crates/ide-assists/src/tests/sourcegen.rs
Comment (jyn514) [inline src/tools/rust-analyzer/crates/ide-assists/src/tests/sourcegen.rs:61]: this seems ... Not great? Why are we ignoring errors? Would like to hear from an RA maintainer that this is ok.

Comment (fasterthanlime) [reply src/tools/rust-analyzer/crates/ide-assists/src/tests/sourcegen.rs:61]: These tests are _really not_ suited to Rust CI.

They:

  * Generate source code
  * Run it through `rustfmt +stable` (not available in CI)
  * Ensure that the version on disk matches the version we just generated

It's great when developing locally, and it even makes sense on their own CI, but we can't run them here. I started by stubbing `reformat` (since we don't have rustfmt stable, and using a stageN rustfmt might give different results), and the tests starting failing because the unformatted version doesn't match the formatted version.

I guess disabling all the tests might be the better option here.. I'll ping RA folks about this.

Comment (jyn514) [reply src/tools/rust-analyzer/crates/ide-assists/src/tests/sourcegen.rs:61]: Ok. I don't mind that they're disabled here. My main concern is you've disabled them in rust-analyzer's CI too.

Would it be helpful if rust-analyzer set a different environment variables, like RUST_ANALYZER_CI=1 or something?

Comment (fasterthanlime) [reply src/tools/rust-analyzer/crates/ide-assists/src/tests/sourcegen.rs]: I ended up reverting those two commits and simply disabling those tests when the `in-rust-tree` feature is enabled.

(The downside is that a _bunch_ of crates have an `in-rust-tree` feature now.. and there's a lot of `cfg` attributes now. Jeeze that's a lot of `cfg` attributes.)

**Comment by jyn514** [comment]
@fasterthanlime I haven't looked at your bootstrap changes in detail, but they look broadly fine (I'll sit down sometime this weekend and either approve or give concrete suggestions). It sounds like you have consensus on this from the RA maintainers which is great - someone on @rust-lang/wg-rls-2 can you confirm?

My main concern is this makes a lot of changes to the rust-analyzer repo *after* the subtree commit, which I would like to have a RA reviewer for. After that I think there aren't any major blockers :)

**Comment by matklad** [comment]
cc @Veykril

**Comment by Veykril** [comment]
I can be the reviewer 👍

**Comment by fasterthanlime** [comment]
Re rust-analyzer changes, here's a summary:

  * Skip slow tests if `SKIP_SLOW_TESTS=1` is set (previously it only checked `CI` and `RUN_SLOW_TESTS`)
  * All sourcegen tests are disabled by cfg gates: that means most ide crates, sourcegen itself
  * Some tidy checks are disabled

Basically, anything that requires a stable rustfmt or a stable cargo, or anything that writes to the source folder needs to be disabled to pass Rust CI. Those offending tests are disabled via the `in-rust-tree` feature of various crates. Enabling `rust-analyzer/in-rust-tree` enables all of them recursively.

The proc-macro-srv tests used to write to the source folder, but that was _very important_ to run in Rust CI, so I landed a fix for that ahead of time: https://github.com/rust-lang/rust-analyzer/pull/12831

The way I've been "simulating Rust CI" locally was:

  * Make source folder read-only (I used `chown -R root:root src/`)
  * Remove all cargo proxies from `$PATH` (I did something like `PATH=$(echo $PATH | tr ':' '\n' | grep -v cargo | tr '\n' ':')`)
  * Run `./x.py test --stage N rust-analyzer` with `CI=1`, `SKIP_SLOW_TESTS=1`

I'm available for any questions!

File: src/tools/rust-analyzer/crates/ide-assists/src/tests/sourcegen.rs
Comment (Veykril) [inline src/tools/rust-analyzer/crates/ide-assists/src/tests/sourcegen.rs:3]: Am I missing something here or why can we not attach the cfg to the `mod` declaration of this `sourcegen` module? Similarly for the `ide-diagnostics` one and the `slow-tests` one. That is I don't see where these differ from the `syntax` crate sourcegen module.

Comment (fasterthanlime) [reply src/tools/rust-analyzer/crates/ide-assists/src/tests/sourcegen.rs:3]: Oh let me try that - I think I just assumed those were [auto-discovered](https://doc.rust-lang.org/rust-by-example/testing/integration_testing.html) - but they're actually in `src`, not `tests`!

Comment (Veykril) [reply src/tools/rust-analyzer/crates/ide-assists/src/tests/sourcegen.rs:3]: Ah okay, ye the sourcegen test in `slow-tests` in the `rust-analyzer` crate are not in the `src` directory, but in `tests`, so that wouldn't really work, but `ide-diagnostics` and `ide-assists` should be good.

Comment (fasterthanlime) [reply src/tools/rust-analyzer/crates/ide-assists/src/tests/sourcegen.rs]: Done in 4962241fd7f75055928fba3cbc4bb817a31266f0

**Review by Veykril — APPROVED** [review]
The changes to rust-analyzer here look good to me

**Comment by tschuett** [comment]
Is it a future option to move the rust-analyzer into the rust repo for tighter integration, i.e., mono repo?

**Comment by fasterthanlime** [comment]
> Is it a future option to move the rust-analyzer into the rust repo for tighter integration, i.e., mono repo?

Rust CI takes 33 minutes to run right now. rust-analyzer CI takes 5 minutes (10 if you count the windows pipeline). There's a big upside to having separate repositories with separate issue trackers, their own CI, etc. The subtree solution prevents future breakage, which is what we want :)

File: src/bootstrap/builder.rs
Comment (Mark-Simulacrum) [inline src/bootstrap/builder.rs:1556]: This looks pretty suspicious to me -- the comment is here for a reason, we shouldn't be expanding this list to arbitrary features, since that means that tools will eventually need cfg(bootstrap) internally, which is painful to support (for us and for the tools). Is there a reason we're not just using Std for rust-analyzer? That seems much more appropriate; it clearly doesn't actually build on stable.

Comment (jyn514) [reply src/bootstrap/builder.rs:1556]: Hmm
Isn't there lots of stuff in bootstrap that already switches on Mode::Std? I worry changing it will break something else.

I do agree the proc macro server is inherently unstable; I guess the best thing to do is to switch and see what breaks. Don't have time for that today.

**Comment by lnicola** [comment]
Perhaps a stupid question, but if we do get a `proc-macro-srv-cli` shipped with the toolchain, doesn't adding the whole RA subtree become less important?

**Comment by fasterthanlime** [comment]
> Perhaps a stupid question, but if we do get a `proc-macro-srv-cli` shipped with the toolchain, doesn't adding the whole RA subtree become less important?

No. Having RA as a subtree is a prerequisite to that. It ensures that the `proc-macro-srv-cli` binary shipped with rustc is _always compatible_ with that rustc. There's no other way to achieve that.

File: src/bootstrap/builder.rs
Comment (jyn514) [reply src/bootstrap/builder.rs:1556]: Ok, having thought about this more I think ToolStd is correct. If you look at when it was originally added, the motivation was to allow distros to build compiletest with the previous beta compiler. Before ToolStd, it would break on change to libtest.

So this does exactly what we want: it allows using nightly features from the standard library, but not language features in the nightly compiler. The fact that it happens to use the beta standard library for build scripts is just because the rustc shim is not very smart and does that unconditionally for all build scripts in stage 0.

Comment (Mark-Simulacrum) [reply src/bootstrap/builder.rs:1556]: proc_macro_span and proc_macro_diagnostic are definitely language-y features, not just library features (even if their surface area is defined there). So I'm not sure I buy that argument fully.

In any case, it seems maybe OK to leave this until it causes us trouble down the line -- as long as rust-analyzer is a subtree it's not too hard to fix breakage caused by these features changing, hopefully.

Comment (jyn514) [reply src/bootstrap/builder.rs:1556]: Sounds good. If it makes you feel better, this allows strictly fewer nightly features than before (it used to be ToolRustc) and the RA devs care a lot about being able to build on stable, so I think this is unlikely to break :)

Comment (Mark-Simulacrum) [reply src/bootstrap/builder.rs:1556]: Well, clearly these features are being used by something that's not going to build on stable, and I'm much more likely to see breakage to them coming from modifications in rustc/std than from work on rust-analyzer, so this is not particularly reassuring :)

ToolRustc is in some ways *better*; it allows all nightly features, but we learn of breakage *immediately*. With ToolStd, we learn of breakage when we do a beta-bump, which is a terrible time. In rust-analyzer's case maybe we're lucky and they're already doing this same thing on their CI and will do a bump to rust-lang/rust in time, but there's a time and someone-is-running-CI-often component there that makes me concerned.

Comment (fasterthanlime) [reply src/bootstrap/builder.rs:1556]: > ToolRustc is in some ways _better_; it allows all nightly features, but we learn of breakage _immediately_

Huh, that might explain why `x.py test --stage 0 rust-analyzer` used to work in my first subtree PR. And it might be the better choice here eventually. 🤔 

Again though, until you think this is a blocker for the merge, for now I'm worried about merge conflicts that would cause me to redo the whole PR (I don't mind doing it if we have a compelling reason, but it is time I'd rather spend elsewhere).

Comment (jyn514) [reply src/bootstrap/builder.rs:1556]: > Huh, that might explain why x.py test --stage 0 rust-analyzer used to work in my first subtree PR. 

That sounds unlikely. The shim uses beta libstd for all stage 0 build scripts regardless of the mode.

I am slightly unsure why @Mark-Simulacrum thinks we'd only catch breakage in beta, since this is only using features from std.

Comment (Mark-Simulacrum) [reply src/bootstrap/builder.rs:1556]: Those features are thin shims over significantly larger surface area in the compiler (AFAIK), like the rest of proc-macro; the compiler used during the use of these features is presumably the stage0 beta, so you're not exercising any new code in the compiler until we bump that beta.

I continue to not believe this is a blocker for merging (otherwise I'd have r-'d), just flagging a concern, partially so that we can get on the same page and make sure @jyn514 is aware of this for future bootstrap changes, potentially unrelated to rust-analyzer. (Or that I'm aware that this is the wrong conclusion to draw, alternatively :)

**Comment by fasterthanlime** [comment]
Just following up on the emoji reactions to [this comment](https://github.com/rust-lang/rust/pull/99603#issuecomment-1193189077): @lnicola and I had a chance to chat about it, so any hesitation there has been hopefully resolved!

@jonas-schievink I've provided more context [in this thread](https://github.com/rust-lang/rust-analyzer/issues/12803#issuecomment-1193192538).

One question I haven't addressed is: "if everything goes according to plan, and proc-macro-srv eventually finds its way out of the rust-analyzer repo, is having RA as a subtree still a good thing?" and the answer is yes, still, for a couple reasons:

  * subtree syncs shows _what actually changed_, not just a couple commit hashes (apart from that, they should be largely the same, just with a different command)
  * this is the preferred model for official rust components (miri is the exception for the time being, due to lack of time). once we started discussing "ra as a subtree", there was a noticeable increase in willingness to put in the work to move this forward
  * it was one of the steps on the path to resolving the whole proc macro issue: for rust & rust-analyzer to share ownership of `proc-macro-srv` for some transition period.

**Comment by jonas-schievink** [comment]
(note that what was asked was "doesn't adding the whole RA subtree become *less important*?", not "doesn't adding the whole RA subtree become *useless*?")

I understood "if we do get a proc-macro-srv-cli shipped with the toolchain" as meaning that rust-lang/rust would, itself, build a tool like that, independently of the rust-analyzer codebase. In that case, the subtree indeed would be unnecessary, hence my confusion.

**Comment by fasterthanlime** [comment]
> I understood "if we do get a proc-macro-srv-cli shipped with the toolchain" as meaning that rust-lang/rust would, itself, build a tool like that, independently of the rust-analyzer codebase. In that case, the subtree indeed would be unnecessary, hence my confusion.

Ah, I see. A proc-macro-srv independent of rust-analyzer's codebase is not in the cards for the short term, no. I believe @lnicola was referring to [this PR](https://github.com/rust-lang/rust-analyzer/pull/12858). Thanks for the clarification!

**Comment by fasterthanlime** [comment]
I recreated the branch & force-pushed due to merge conflicts, but reading https://github.com/rust-lang/rust/pull/99603#issuecomment-1193258804, it has some.. fishy messages?

```
Adding src/tools/rust-analyzer/crates/vfs/src/lib.rs
Addingerror: add_cacheinfo failed to refresh for path 'src/tools/rust-analyzer/crates/syntax/test_data/parser/validation/0046_mutable_const_item.rs'; merge aborting.
 src/tools/rust-analyzer/crates/vfs/src/file_set/tests.rs
```

**Comment by lnicola** [comment]
Maybe we could at least do a @bors try until it gets approved again?

**Comment by fasterthanlime** [comment]
The failing tests rely on `FxHashSet` ordering, which isn't the same on 64-bit and 32-bit, thus the tests fail on i686:

```
error: expect test failed
--> crates/hir-def/src/import_map.rs:862:13
You can update all `expect!` tests by running:
    env UPDATE_EXPECT=1 cargo test
To update a single test, place the cursor on `expect` token and use `run` feature of rust-analyzer.
Expect:
----
dep::fmt::Display::format_method (a)
dep::fmt::Display::FMT_CONST (a)
dep::fmt::Display::format_function (a)
----
Actual:
----
dep::fmt::Display::format_method (a)
dep::fmt::Display::format_function (a)
dep::fmt::Display::FMT_CONST (a)
```

**Comment by jyn514** [comment]
@fasterthanlime you could try using BTreeMap instead, it uses Ord instead of Hash, which should be stable across platforms

**Comment by jyn514** [comment]
@lnicola reports 2 tests still failing: https://rust-lang.zulipchat.com/#narrow/stream/185405-t-compiler.2Frust-analyzer/topic/git.20subtree.3A.20last.20stretch

@bors r-

**Pull request merged**
